src/software/thunderscope/gl/layers/gl_max_dribble_layer.py

```diff
@@ -0,0 +1,76 @@
+from pyqtgraph.opengl import *
+
+from software.thunderscope.constants import Colors, DepthValues
+from software.thunderscope.thread_safe_buffer import ThreadSafeBuffer
+from software.thunderscope.gl.layers.gl_layer import GLLayer
+from software.thunderscope.gl.graphics.gl_circle import GLCircle
+from software.thunderscope.gl.graphics.gl_polygon import GLPolygon
+import software.python_bindings as tbots_cpp
+from proto.import_all_protos import *
+
+from pyqtgraph.Qt import QtGui
+
+from software.thunderscope.gl.helpers.observable_list import ObservableList
+
+import math
+
+class GLMaxDribbleLayer(GLLayer):
+    """GLLayer"""
+
+    def __init__(self, name: str, buffer_size:int = 5) -> None:
+        """Initialize the GLMaxDribbleLayer
+
+        :param name: The displayed name of the layer
+        :param buffer_size: The buffer size, set higher for smoother plots.
+                            Set lower for more realtime plots. Default is arbitrary
+                            """
+        super().__init__(name)
+        self.setDepthValue(DepthValues.BACKGROUND_DEPTH)
+
+        self.world_buffer = ThreadSafeBuffer(buffer_size, World)
+
+        self.color = QtGui.QColor(255, 0, 0, 127)
+
+        self.dribble_radius_graphic = GLPolygon(parent_item = self, outline_color = self.color)
+        self.dribble_circle_graphic = GLCircle(parent_item = self, outline_color = self.color)
+        self.dribble_circle_graphic.set_radius(1, 24)
+
+    def refresh_graphics(self) -> None:
+        """Update graphics in this layer"""
+
+        world = self.world_buffer.get(block=False)
+        if world.HasField("dribble_displacement"):
+            dribbleDisp = world.dribble_displacement
+
+            dist = ((dribbleDisp.start.x_meters - dribbleDisp.end.x_meters)**2 + (dribbleDisp.start.y_meters - dribbleDisp.end.y_meters)**2)**(1/2)
```

Comment:
```suggestion
            dist = tbots_cpp.createSegment(dribble_disp).length()
```

We have binded some useful C++ functions to make this code simpler.
- `createSegment` ([C++](https://github.com/UBC-Thunderbots/Software/blob/81ec0830e11d0a53002aa517a2cc190de75a9721/src/proto/message_translation/tbots_geometry.cpp#L110)) ([pybind](https://github.com/UBC-Thunderbots/Software/blob/81ec0830e11d0a53002aa517a2cc190de75a9721/src/software/python_bindings.cpp#L300))
- `Segment::length` ([C++](https://github.com/UBC-Thunderbots/Software/blob/81ec0830e11d0a53002aa517a2cc190de75a9721/src/software/geom/segment.h#L52)) ([pybind](https://github.com/UBC-Thunderbots/Software/blob/81ec0830e11d0a53002aa517a2cc190de75a9721/src/software/python_bindings.cpp#L243))

Reply:
I was trying to do this without the createSegment method, since I'm pretty sure dribble_disp should already be a segment. Can you explain why we have to cast it to segment before we can call length()? (Something to do with protobuffer?)

Reply:
Good question.

Your `dribble_disp` variable is taken from the `World` proto: https://github.com/UBC-Thunderbots/Software/blob/0730446a96e982cc9b302c8396fbf6a60c9c9861/src/proto/world.proto#L33. This `Segment` is a protobuf type: https://github.com/UBC-Thunderbots/Software/blob/0730446a96e982cc9b302c8396fbf6a60c9c9861/src/proto/geometry.proto#L54.

This is different from our C++ `Segment`. The protobuf compiler will automatically take the `.proto` file and generate some C++ and Python classes, so this is a protobuf `Segment` class that contains `start` and `end` members.

Our `Segment` class is different from the protobuf one and [has more sugar](https://github.com/UBC-Thunderbots/Software/blob/0730446a96e982cc9b302c8396fbf6a60c9c9861/src/software/geom/segment.h#L6). It defines a whole bunch of useful helpers that we could use.

So, we've defined translation functions between the `Segment` proto class and the `Segment` Thunderbots class: `https://github.com/UBC-Thunderbots/Software/blob/0730446a96e982cc9b302c8396fbf6a60c9c9861/src/proto/message_translation/tbots_geometry.h#L43`. This is why calling `createSegment` allows us to call the `length()` helper.